plugins/GigPlayer/GigPlayer.cpp

```diff
@@ -333,89 +339,73 @@ void GigInstrument::play( SampleFrame* _working_buffer )
+				else
 				{
-					sample.adsr.keyup();
-				}
+					gigNote.state = GigState::PlayingKeyUp;
 
-				// Add release samples if available
-				if( it->release == true )
-				{
-					addSamples( *it, true );
+					// Notify each sample that the key has been released
+					for (auto& sample : gigNote.samples) { sample.adsr.keyup(); }
+
+					// Add release samples if available
+					if (gigNote.release) { addSamples(gigNote, true); }
 				}
 			}
-		}
-		// Process notes in the KeyDown state, adding samples for the notes
-		else if( it->state == GigState::KeyDown )
-		{
-			it->state = GigState::PlayingKeyDown;
-			addSamples( *it, false );
-		}
-
-		// Delete ended samples
-		for( QList<GigSample>::iterator sample = it->samples.begin();
-				sample != it->samples.end(); ++sample )
-		{
-			// Delete if the ADSR for a sample is complete for normal
-			// notes, or if a release sample, then if we've reached
-			// the end of the sample
-			if( sample->sample == nullptr || sample->adsr.done() ||
-				( it->isRelease == true &&
-				  sample->pos >= sample->sample->SamplesTotal - 1 ) )
+			// Process notes in the KeyDown state, adding samples for the notes
+			else if (gigNote.state == GigState::KeyDown)
 			{
-				sample = it->samples.erase( sample );
-
-				if( sample == it->samples.end() )
-				{
-					break;
-				}
+				gigNote.state = GigState::PlayingKeyDown;
+				addSamples(gigNote, false);
 			}
-		}
 
-		// Delete ended notes (either in the completed state or all the samples ended)
-		if( it->state == GigState::Completed || it->samples.empty() )
+			// TODO: C++20 std::erase_if
+			// Delete ended samples
+			gigNote.samples.erase(std::remove_if(gigNote.samples.begin(), gigNote.samples.end(),
+				[&](GigSample& sample) {
+					if (sample.adsr.done()) { return true; }
+					if (sample.sample == nullptr) { return true; }
+					if (gigNote.isRelease && sample.pos >= (sample.sample->SamplesTotal - 1))
+					{
+						return true;
+					}
+					return false;
+				}),
+				gigNote.samples.end()
+			);
+
+			// Delete ended notes (either in the completed state or all the samples ended)
+			return gigNote.state == GigState::Completed || gigNote.samples.empty();
+		}),
+		m_notes.end()
+	);
+
+	auto updateWorkingBufferByData = [&](const auto& data) {
+		for (f_cnt_t i = 0; i < frames; ++i)
 		{
-			it = m_notes.erase( it );
-
-			if( it == m_notes.end() )
-			{
-				break;
-			}
+			_working_buffer[i][0] += data[i][0];
+			_working_buffer[i][1] += data[i][1];
 		}
-	}
+	};
 
 	// Fill buffer with portions of the note samples
-	for (auto& note : m_notes)
-	{
-		// Only process the notes if we're in a playing state
-		if (!(note.state == GigState::PlayingKeyDown || note.state == GigState::PlayingKeyUp ))
-		{
-			continue;
-		}
+	std::for_each(m_notes.begin(), m_notes.end(), [&](GigNote& note) {
+		if (!(note.state == GigState::PlayingKeyDown || note.state == GigState::PlayingKeyUp)) { return; }
```

Comment:
```suggestion
		if (note.state != GigState::PlayingKeyDown && note.state != GigState::PlayingKeyUp) { return; }
```
Simplified using De Morgan's laws